### ISSUE TITLE
mapTo + flatten might need lazy evaluation

### ISSUE THREAD

**Issue by staltz** [staltz/xstream#35]
There are cases where mapTo(innerStream) is simply not the same as `map(() => innerStream)` because the former does lazy evaluation of the `innerStream` expression, which may involve the creation of a new stream, eg if `innerStream` expression was `source.compose(dropUntil(second))`.

Here's an example:

```diff
 function notBetween(first, second) {
   return source => xs.merge(
     source.endWhen(first),
-    first.mapTo(source.compose(dropUntil(second))).flatten()
+    first.map(() => source.compose(dropUntil(second))).flatten()
   )
 }
```

The `map` works but not the `mapTo`.

This bug/behavior is not a top priority, but something we should be aware of, in case someone else trips over this.

**Comment by TylorS** [comment]
I think that this is solved now since `mapTo(x)` is really just a call to `map(() => x)`

**Comment by staltz** [comment]
Not really, because `x` is eagerly evaluated in `mapTo(x)` but lazily evaluated in `map(() => x)`. This issue is marked bug but I think it's rather unsolvable. What we could do is forbid passing a stream to mapTo, so we always pass streams in map, for lazy evaluation.

**Comment by TylorS** [comment]
But they are the same now, no? https://github.com/staltz/xstream/blob/master/src/core.ts#L1699

**Comment by staltz** [comment]
Not really, because

```js
function multiply(x) {
  return x * Math.random();
}
```

is different to
```js
const multiply = makeMultiplyFunction(Math.random());
```

given
```js
function makeMultiplyFunction(c) {
  return function multiply(x) {
    return x * c;
  } 
}
```
